traitsui/tests/editors/test_image_enum_editor.py

```diff
@@ -0,0 +1,501 @@
+class TestSimpleImageEnumEditor(unittest.TestCase):
+
+    def setup_gui(self, model, view):
+        gui = GUI()
+        ui = model.edit_traits(view=view)
+        self.addCleanup(ui.dispose)
+
+        gui.process_events()
+        editor = ui.get_editors("value")[0]
+        control = editor.control
+
+        return gui, control
+
+    def test_simple_editor_more_cols(self):
+        # Smoke test for setting up an editor with more than one column
+        enum_edit = EnumModel()
+        view = View(
+            UItem(
+                'value',
+                editor=ImageEnumEditor(
+                    values=[
+                        'top left', 'top right', 'bottom left', 'bottom right'
+                    ],
+                    prefix='@icons:',
+                    suffix='_origin',
+                    path='dummy_path',
+                    cols=4,
+                ),
+                style="simple",
+            ),
+            resizable=True,
+        )
+
+        with store_exceptions_on_all_threads():
+            self.setup_gui(enum_edit, view)
+
+    @skip_if_not_wx
+    def test_simple_editor_popup_editor(self):
+        enum_edit = EnumModel()
+
+        with store_exceptions_on_all_threads():
+            with patch(cache_to_patch, wraps=image_cache) as patched_cache:
+                gui, control = self.setup_gui(enum_edit, get_view("simple"))
+
+                self.assertEqual(enum_edit.value, 'top left')
+                self.assertEqual(
+                    patched_cache.call_args[0][0], "@icons:top left_origin"
+                )
+
+                # Set up ImageEnumDialog
+                click_on_image(control)
+                gui.process_events()
+
+                # Check created buttons
+                image_grid_control = control.GetChildren()[0].GetChildren()[0]
+                self.assertEqual(
+                    get_button_strings(image_grid_control),
+                    ['top left', 'top right', 'bottom left', 'bottom right']
+                )
+
+                # Select new image
+                click_on_image(get_button_control(image_grid_control, 1))
+                gui.process_events()
+
+                self.assertEqual(enum_edit.value, 'top right')
+                self.assertEqual(
+                    patched_cache.call_args[0][0], "@icons:top right_origin"
+                )
+
+                # Check that dialog window is closed
+                self.assertEqual(list(control.GetChildren()), [])
+
+    @skip_if_not_qt4
+    def test_simple_editor_combobox(self):
+        enum_edit = EnumModel()
+
+        with store_exceptions_on_all_threads():
+            with patch(cache_to_patch, wraps=image_cache) as patched_cache:
+                gui, combobox = self.setup_gui(enum_edit, get_view("simple"))
+
+                self.assertEqual(enum_edit.value, 'top left')
+                self.assertEqual(
+                    patched_cache.call_args[0][0], "@icons:top left_origin"
```

Comment:
Yesterday I ran these tests with a Linux machine of mine with PyQt5 and a few of these tests checking the last call argument on the mock were consistently failing ("@icon:bottom right_origin" != "@icons: top left_origin". On close inspection, all the images were fetched from the cache, but the painted one is not necessarily the first one, the GUI itself was behaving fine. I haven't seen the same failures on Travis/Appveyor or my own OSX machine.

I understand this is an indirect test on what image ended up being used, and there is otherwise no way to check for this. Unfortunately this assertion might just be a bit too indirect and inconclusive, and is depending too strongly on implementation details that may vary from platforms to platforms, from one Qt build to another, causing the test to be unreliable. I think we shall accept that there is no reasonable way to inspect the image being used as of now and remove these checks.  The assertion on `enum_edit.value` is good and can be kept.

Reply:
I removed the checks from `TestSimpleImageEnumEditor` but not from `TestReadOnlyImageEnumEditor`. For readonly editors that is the only reasonable way to test them, and even though those assertions rely on implementation details, they shouldn't fail if `update_editor` is called as expected.

Reply:
Given the readonly editor only shows one image: the selected one, that makes sense. Thank you.